File: iOSCategories/NSDate+Additions.m
//
//  NSDate+Additions.m
//
//  Created by Applico Inc. on 7/22/14.
//  Copyright (c) 2014 Applico Inc. All rights reserved.
//

#import "NSDate+Additions.h"

@implementation NSDate (Additions)

- (NSDate *)previousDay {
    NSDate *date = nil;
    NSCalendar *calendar = [NSCalendar currentCalendar];
    NSDateComponents *dateComponents = [calendar components:(NSMonthCalendarUnit | NSDayCalendarUnit | NSYearCalendarUnit | NSCalendarUnitHour | NSMinuteCalendarUnit) fromDate:self];
    
    [dateComponents setDay:[dateComponents day]-1];
    
    date = [calendar dateFromComponents:dateComponents];
    
    return date;
    
}

- (NSDate *)previousDayWithoutTime {
    NSDate *date = nil;
    NSCalendar *calendar = [NSCalendar currentCalendar];
    NSDateComponents *dateComponents = [calendar components:(NSMonthCalendarUnit | NSDayCalendarUnit | NSYearCalendarUnit | NSHourCalendarUnit | NSMinuteCalendarUnit) fromDate:self];
    
    [dateComponents setDay:[dateComponents day]-1];
    [dateComponents setHour:0];
    [dateComponents setMinute:0];
    [dateComponents setSecond:0];
    
    date = [calendar dateFromComponents:dateComponents];
    
    return date;
    
}

- (NSDate *)dateWithoutTime {
    NSDate *date = nil;
    NSCalendar *calendar = [NSCalendar currentCalendar];
    NSDateComponents *dateComponents = [calendar components:(NSMonthCalendarUnit | NSDayCalendarUnit | NSYearCalendarUnit | NSHourCalendarUnit | NSMinuteCalendarUnit) fromDate:self];
    
    [dateComponents setHour:0];
    [dateComponents setMinute:0];
    [dateComponents setSecond:0];
    
    date = [calendar dateFromComponents:dateComponents];
    
    return date;
    
}

- (NSString *)shortStyleTimeString {
    
    NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
    [dateFormatter setDateStyle:NSDateFormatterNoStyle];
    [dateFormatter setTimeStyle:NSDateFormatterShortStyle];
    
    NSTimeZone *timezone = [NSTimeZone timeZoneWithAbbreviation:@"UTC"];
    [dateFormatter setTimeZone:timezone];

    return [dateFormatter stringFromDate:self];
    
}

- (NSString *)shortStyleString {
    NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
    [dateFormatter setDateStyle:NSDateFormatterShortStyle];
    [dateFormatter setTimeStyle:NSDateFormatterShortStyle];
    NSTimeZone *timezone = [NSTimeZone timeZoneWithAbbreviation:@"UTC"];
    [dateFormatter setTimeZone:timezone];
    
    return [dateFormatter stringFromDate:self];
    
}

- (NSString *)longStyleString {
    NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
    [dateFormatter setDateStyle:NSDateFormatterLongStyle];
    [dateFormatter setTimeStyle:NSDateFormatterLongStyle];
    
    NSTimeZone *timezone = [NSTimeZone timeZoneWithAbbreviation:@"UTC"];
    [dateFormatter setTimeZone:timezone];
    
    return [dateFormatter stringFromDate:self];
    
}

- (NSDate *)dateWithUTCTimeZone {
    NSDate *date = nil;
    NSCalendar *calendar = [NSCalendar currentCalendar];

    NSDateComponents *dateComponents = [calendar components:(NSMonthCalendarUnit | NSDayCalendarUnit | NSYearCalendarUnit | NSHourCalendarUnit | NSMinuteCalendarUnit) fromDate:self];
    
    [calendar setTimeZone:[NSTimeZone localTimeZone]];
    
    date = [calendar dateFromComponents:dateComponents];
    
    return date;
    
}

- (NSString *)daysAndHoursSinceDate:(NSDate *)date {

    NSMutableString *returnString = [[NSMutableString alloc] init];
    NSCalendar *calendar = [NSCalendar currentCalendar];
    [calendar setTimeZone:[NSTimeZone localTimeZone]];
    
    NSDateComponents *components = [calendar components:(NSMonthCalendarUnit | NSDayCalendarUnit | NSYearCalendarUnit | NSHourCalendarUnit | NSMinuteCalendarUnit)
                                                fromDate:date
                                                  toDate:[NSDate date] options:0];
    
    
    if([components month] > 0) {
        if([components month] > 1) {
            [returnString appendString:@"s"];
        }
        [returnString appendFormat:@"%dM ", [components month]];
       
    }
    
    if([components day] > 0) {
        if([components day] > 1) {
            [returnString appendString:@"s"];
        }
        [returnString appendFormat:@"%dd ", [components day]];

    }
    
    if([components hour] > 0) {
        if([components hour] > 1) {
            [returnString appendString:@"s"];
        }
        [returnString appendFormat:@"%dh ", [components hour]];

    }
    
    if([components minute] > 0) {
        if([components minute] > 1) {
            [returnString appendString:@"s"];
        }
        [returnString appendFormat:@"%d min", [components minute]];

    }
    

    if([returnString length] > 0) {
        [returnString appendString:@" ago"];
        return returnString;
    }

    return @"now";
}

@end

File: Swift/NSDate+Additions.swift
//
//  NSDate+Additions.swift
//  99Bottles-Swift
//
//  Created by Josh L on 6/2/15.
//  Copyright (c) 2015 Applico. All rights reserved.
//

import Foundation

extension NSDate {
    
    func previousDay() -> NSDate? {
        
        let calendar = NSCalendar.currentCalendar()
        let dateComponents = calendar.components(NSCalendarUnit.CalendarUnitMonth | NSCalendarUnit.CalendarUnitDay | NSCalendarUnit.CalendarUnitYear, fromDate: self)
        
        dateComponents.day = dateComponents.day - 1
        
        return calendar.dateFromComponents(dateComponents)
    }
    
    func previousDayWithoutTime() -> NSDate? {
        
        let calendar = NSCalendar.currentCalendar()
        let dateComponents = calendar.components(NSCalendarUnit.CalendarUnitMonth | NSCalendarUnit.CalendarUnitDay | NSCalendarUnit.CalendarUnitYear, fromDate: self)
        
        dateComponents.day = dateComponents.day - 1
        dateComponents.hour = 0
        dateComponents.minute = 0
        dateComponents.second = 0
        
        return calendar.dateFromComponents(dateComponents)
    }
    
    func dateWithoutTime() -> NSDate? {
        
        let calendar = NSCalendar.currentCalendar()
        let dateComponents = calendar.components(NSCalendarUnit.CalendarUnitMonth | NSCalendarUnit.CalendarUnitDay | NSCalendarUnit.CalendarUnitYear, fromDate: self)

        dateComponents.hour = 0
        dateComponents.minute = 0
        dateComponents.second = 0
        
        return calendar.dateFromComponents(dateComponents)
    }
    
    func shortStyleTimeString() -> String {
        
        let dateFormatter = NSDateFormatter()
        dateFormatter.dateStyle = NSDateFormatterStyle.NoStyle
        dateFormatter.timeStyle = NSDateFormatterStyle.ShortStyle
        dateFormatter.timeZone = NSTimeZone(abbreviation: "UTC")
        
        return dateFormatter.stringFromDate(self)
    }
    
    func shortStyleString() -> String {
        
        let dateFormatter = NSDateFormatter()
        dateFormatter.dateStyle = NSDateFormatterStyle.ShortStyle
        dateFormatter.timeStyle = NSDateFormatterStyle.ShortStyle
        dateFormatter.timeZone = NSTimeZone(abbreviation: "UTC")
        
        return dateFormatter.stringFromDate(self)
    }
    
    func longStyleString() -> String {
        
        let dateFormatter = NSDateFormatter()
        dateFormatter.dateStyle = NSDateFormatterStyle.LongStyle
        dateFormatter.timeStyle = NSDateFormatterStyle.LongStyle
        dateFormatter.timeZone = NSTimeZone(abbreviation: "UTC")
        
        return dateFormatter.stringFromDate(self)
    }
    
    func dateWithUTCTimeZone() -> NSDate? {
        
        let calendar = NSCalendar.currentCalendar()
        let dateComponents = calendar.components(NSCalendarUnit.CalendarUnitMonth | NSCalendarUnit.CalendarUnitDay | NSCalendarUnit.CalendarUnitYear, fromDate: self)
        
        calendar.timeZone = NSTimeZone.localTimeZone()
        
        return calendar.dateFromComponents(dateComponents)
    }
    
    func daysAndHoursSinceDate() -> String {
        
        var returnString = String()
        let calendar = NSCalendar.currentCalendar()
        calendar.timeZone = NSTimeZone.localTimeZone()
        
        let components = calendar.components(NSCalendarUnit.CalendarUnitMonth | NSCalendarUnit.CalendarUnitDay | NSCalendarUnit.CalendarUnitYear | NSCalendarUnit.CalendarUnitHour | NSCalendarUnit.CalendarUnitMinute, fromDate: self, toDate: NSDate(), options: NSCalendarOptions(0))
        
        if components.month > 0 {
            if components.month > 1 {
                returnString = returnString + "s"
            }
            returnString = returnString + String(format: "%ldM", components.month)
        }
        
        if components.day > 0 {
            if components.day > 1 {
                returnString = returnString + "s"
            }
            returnString = returnString + String(format: "%ldd", components.day)
        }
        
        if components.hour > 0 {
            if components.hour > 1 {
                returnString = returnString + "s"
            }
            returnString = returnString + String(format: "%ldh", components.hour)
        }
        
        if components.minute > 0 {
            if components.minute > 1 {
                returnString = returnString + "s"
            }
            returnString = returnString + String(format: "%ld min", components.minute)
        }
        
        if !returnString.isEmpty {
            return returnString + " ago"
        }
        
        return "now"
    }
}